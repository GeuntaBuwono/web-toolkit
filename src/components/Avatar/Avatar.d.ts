type AvatarSize = 'xs' | 'sm' | 'md' | 'lg';

type AvatarType = {
	name: string;
	src?: string;
	size?: AvatarSize;
	alt?: string;
};
