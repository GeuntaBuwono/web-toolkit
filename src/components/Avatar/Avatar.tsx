import getInitialName from '@utils/getInitialName';

export type AvatarSize = 'xs' | 'sm' | 'md' | 'lg';

export type AvatarProps = {
	name: string;
	src?: string;
	size?: AvatarSize;
	alt?: string;
};

const avatarSizeMapper = ({size}: {size: AvatarSize}) => {
	switch (size) {
		case 'xs':
			return 'h-6 w-6';
		case 'sm':
			return 'h-8 w-8';
		case 'md':
			return 'h-10 w-10';
		case 'lg':
			return 'h-12 w-12';
	}
};

const avatarTextSizeMapper = ({size}: {size: AvatarSize}) => {
	switch (size) {
		case 'xs':
			return 'text-[0.5rem] p-1';
		case 'sm':
			return 'text-sm p-1.5';
		case 'md':
			return 'text-base p-2';
		case 'lg':
			return 'text-lg p-3';
	}
};

const Avatar = ({size = 'md', alt = '', src = '', name}: AvatarProps) => {
	if (src) {
		return (
			<img
				className={`inline-block rounded-full ring-2 ring-white ${avatarSizeMapper(
					{
						size,
					},
				)}`}
				src={src}
				alt={alt}
			/>
		);
	}
	return (
		<p
			className={`
			${avatarTextSizeMapper({size})}
			${avatarSizeMapper({size})}
			rounded-full border-2 flex relative bg-gray-300 text-gray-500 items-center`}
		>
			{getInitialName({name})}
		</p>
	);
};

export default Avatar;
