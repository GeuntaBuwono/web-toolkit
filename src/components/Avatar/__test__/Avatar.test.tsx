/**
 * @jest-environment jsdom
 */

import Avatar from '@components/Avatar/Avatar';
import {render} from '@testing-library/react';

const expectedTestValue = {
	src: 'https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/170.jpg',
	alt: 'The image alt tag for the default Avatar Component',
	name: 'Tomi Budi LastName',
};

describe('Avatar', () => {
	test('renders default', () => {
		const {getByText} = render(<Avatar name={expectedTestValue.name} />);
		const testAvatarElement = document.querySelector('p') as HTMLImageElement;

		expect(testAvatarElement).toMatchSnapshot();
		expect(testAvatarElement?.src).toBeUndefined();
		expect(testAvatarElement?.alt).toBeUndefined();

		expect(getByText('TL')).toBeTruthy();
		expect(testAvatarElement.className).toContain('h-10 w-10');
	});

	test('renders with size xs', () => {
		const {getByText} = render(
			<Avatar name={expectedTestValue.name} size="xs" />,
		);
		const testAvatarElement = document.querySelector('p') as HTMLImageElement;

		expect(testAvatarElement).toMatchSnapshot();
		expect(testAvatarElement?.src).toBeUndefined();
		expect(testAvatarElement?.alt).toBeUndefined();

		expect(getByText('TL')).toBeTruthy();
		expect(testAvatarElement.className).toContain('h-6 w-6');
	});

	test('renders with size sm', () => {
		const {getByText} = render(
			<Avatar name={expectedTestValue.name} size="sm" />,
		);
		const testAvatarElement = document.querySelector('p') as HTMLImageElement;

		expect(testAvatarElement).toMatchSnapshot();
		expect(testAvatarElement?.src).toBeUndefined();
		expect(testAvatarElement?.alt).toBeUndefined();

		expect(getByText('TL')).toBeTruthy();
		expect(testAvatarElement.className).toContain('h-8 w-8');
	});

	test('renders with size md', () => {
		const {getByText} = render(
			<Avatar name={expectedTestValue.name} size="md" />,
		);
		const testAvatarElement = document.querySelector('p') as HTMLImageElement;

		expect(testAvatarElement).toMatchSnapshot();
		expect(testAvatarElement?.src).toBeUndefined();
		expect(testAvatarElement?.alt).toBeUndefined();

		expect(getByText('TL')).toBeTruthy();
		expect(testAvatarElement.className).toContain('h-10 w-10');
	});

	test('renders with size lg', () => {
		const {getByText} = render(
			<Avatar name={expectedTestValue.name} size="lg" />,
		);
		const testAvatarElement = document.querySelector('p') as HTMLImageElement;

		expect(testAvatarElement).toMatchSnapshot();
		expect(testAvatarElement?.src).toBeUndefined();
		expect(testAvatarElement?.alt).toBeUndefined();

		expect(getByText('TL')).toBeTruthy();
		expect(testAvatarElement.className).toContain('h-12 w-12');
	});

	test('renders with src size default', () => {
		render(<Avatar {...expectedTestValue} />);
		const testAvatarElement = document.querySelector('img') as HTMLImageElement;

		expect(testAvatarElement).toMatchSnapshot();
		expect(testAvatarElement.src).toContain(expectedTestValue.src);
		expect(testAvatarElement.alt).toContain(expectedTestValue.alt);
		expect(testAvatarElement.className).toContain('h-10 w-10');
	});

	test('renders with xs size', () => {
		render(<Avatar {...expectedTestValue} size="xs" />);
		const testAvatarElement = document.querySelector('img') as HTMLImageElement;

		expect(testAvatarElement).toMatchSnapshot();
		expect(testAvatarElement.src).toContain(expectedTestValue.src);
		expect(testAvatarElement.alt).toContain(expectedTestValue.alt);
		expect(testAvatarElement.className).toContain('h-6 w-6');
	});

	test('renders with sm size', () => {
		render(<Avatar {...expectedTestValue} size="sm" />);
		const testAvatarElement = document.querySelector('img') as HTMLImageElement;

		expect(testAvatarElement).toMatchSnapshot();
		expect(testAvatarElement.src).toContain(expectedTestValue.src);
		expect(testAvatarElement.alt).toContain(expectedTestValue.alt);
		expect(testAvatarElement.className).toContain('h-8 w-8');
	});

	test('renders with md size', () => {
		render(<Avatar {...expectedTestValue} size="md" />);
		const testAvatarElement = document.querySelector('img') as HTMLImageElement;

		expect(testAvatarElement).toMatchSnapshot();
		expect(testAvatarElement.src).toContain(expectedTestValue.src);
		expect(testAvatarElement.alt).toContain(expectedTestValue.alt);
		expect(testAvatarElement.className).toContain('h-10 w-10');
	});

	test('renders with lg size', () => {
		render(<Avatar {...expectedTestValue} size="lg" />);
		const testAvatarElement = document.querySelector('img') as HTMLImageElement;

		expect(testAvatarElement).toMatchSnapshot();
		expect(testAvatarElement.src).toContain(expectedTestValue.src);
		expect(testAvatarElement.alt).toContain(expectedTestValue.alt);
		expect(testAvatarElement.className).toContain('h-12 w-12');
	});
});
