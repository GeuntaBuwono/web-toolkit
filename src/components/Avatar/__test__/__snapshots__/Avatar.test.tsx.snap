// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar renders default 1`] = `
<p
  class="
			text-base p-2
			h-10 w-10
			rounded-full border-2 flex relative bg-gray-300 text-gray-500 items-center"
>
  TL
</p>
`;

exports[`Avatar renders with lg size 1`] = `
<img
  alt="The image alt tag for the default Avatar Component"
  class="inline-block rounded-full ring-2 ring-white h-12 w-12"
  src="https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/170.jpg"
/>
`;

exports[`Avatar renders with md size 1`] = `
<img
  alt="The image alt tag for the default Avatar Component"
  class="inline-block rounded-full ring-2 ring-white h-10 w-10"
  src="https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/170.jpg"
/>
`;

exports[`Avatar renders with size lg 1`] = `
<p
  class="
			text-lg p-3
			h-12 w-12
			rounded-full border-2 flex relative bg-gray-300 text-gray-500 items-center"
>
  TL
</p>
`;

exports[`Avatar renders with size md 1`] = `
<p
  class="
			text-base p-2
			h-10 w-10
			rounded-full border-2 flex relative bg-gray-300 text-gray-500 items-center"
>
  TL
</p>
`;

exports[`Avatar renders with size sm 1`] = `
<p
  class="
			text-sm p-1.5
			h-8 w-8
			rounded-full border-2 flex relative bg-gray-300 text-gray-500 items-center"
>
  TL
</p>
`;

exports[`Avatar renders with size xs 1`] = `
<p
  class="
			text-[0.5rem] p-1
			h-6 w-6
			rounded-full border-2 flex relative bg-gray-300 text-gray-500 items-center"
>
  TL
</p>
`;

exports[`Avatar renders with sm size 1`] = `
<img
  alt="The image alt tag for the default Avatar Component"
  class="inline-block rounded-full ring-2 ring-white h-8 w-8"
  src="https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/170.jpg"
/>
`;

exports[`Avatar renders with src size default 1`] = `
<img
  alt="The image alt tag for the default Avatar Component"
  class="inline-block rounded-full ring-2 ring-white h-10 w-10"
  src="https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/170.jpg"
/>
`;

exports[`Avatar renders with xs size 1`] = `
<img
  alt="The image alt tag for the default Avatar Component"
  class="inline-block rounded-full ring-2 ring-white h-6 w-6"
  src="https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/170.jpg"
/>
`;
